### DISCUSSION TITLE
Working Through This

### DISCUSSION THREAD

**Discussion by heffneil** [ballle98/AqualinkD#62] [General]
Can we chat here so I can learn more about AqualinkD and PDA's and programming?  

My PDA is wireless (obviously) and it's very very very slow (is that normal).  My previous home had a RS6 keypad hardwired and it worked pretty well.

What I do is since we have freeze protect and I leave my pool open during the winter I run my 8 hours of cleaning overnight so that if the daytime weather isn't very cold, but the night time is I won't run the pump for 16+ hours but the 8 hours overlap with freeze protect (does that make sense?). What I did before I got Aqualinkd working with updating the time was just changing the panel hours so it ran the cleaning cycle during the day.  Sigh - so now I need to update my programs to work during the daytime hours.

I thought about using HA to control the systems but it's wifi and I don't know if I trust it to be able to connect to turn things off.

**Comment by ballle98** [comment]
In my setup I have a 30 ft RS485 connection using CAT5 cable that runs up the wall into the attic and then into the house.  My pi is inside the house and has a wired ethernet connection so no wi-fi issues.  If you have a newer pi you can run openhab and aqualinkd on the same device.  Then you can script everything in openhab.  You could even have it change things based on weather.

It is possible to add the program functions to aqualinkd.  They are pretty buried in the PDA menus but so was time and freeze protect. 
The part that I'm no good at is the web UI.  If Shaun adds support to RS panel I can make it work for PDA.

**Reply by heffneil** [reply]
This sounds like a smart setup.  I don't know what location you are in but i had a lot of transient lightning surge issues in Florida.  I went to an RS485 Opto isolator to save my stuff from exploding - or something fiber?  I am not sure I have this option unfortunately but I will look at other options.

**Comment by heffneil** [comment]
wouldn't a portion of the code be in that other referenced project?  Just curious if there's a way to incorporate it and have a fantasy web interface.  I have considered replacing the chip on my board and even talked to the guy who sells the parts but I don't have a non PDA panel anywhere or a way to really get it anywhere.  The wireless PDA is slow and frustrating.

I remember people (pool folks) coming to my house and plugging in their own RS interface right at the panel to make changes and check stuff.  I could put one in a box outside but that doesn't seem better.

**Reply by ballle98** [reply]
https://github.com/earlephilhower/aquaweb is just a dumb terminal.  Like a screen and a keyboard.  The PDA is simple it just displays what the control panel tells it to display and sends key presses to the control panel to process.  AqualinkD is doing a translation from a press on the web GUI to simulating the PDA actions of 
1. select to wake 
2. wait for home menu
3. press up to select equipment menu and select
4. press up or down to highligh item correstonding to equipment to turn on/off
5. press select to toggle item

**Comment by heffneil** [comment]
So this morning - the system is showing the pump running - but its not.  I stopped the service. moved the log file out and restarted the service.  When doing that is properly shows the pump as NOT running.  For whatever reason, it is hung up on an OLD status.

**Comment by heffneil** [comment]
This morning it still shows it running - but its not running :(

Debug below:

May-04-22 08:06:13 AM Debug:   NetService:Served WEB request
May-04-22 08:06:13 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:13 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:13 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:13 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:14 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:14 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:14 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:14 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:14 AM Info:    NetService:URI request: '/api/debug/status '
May-04-22 08:06:14 AM Debug:   NetService:API: URI Request 'debug/status': value 0.00
May-04-22 08:06:14 AM Debug:   NetService:Served WEB request
May-04-22 08:06:14 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:14 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:14 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:14 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:15 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:15 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:15 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:15 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:15 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:15 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:15 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:15 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:16 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:16 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:16 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:16 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:16 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:16 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:16 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:16 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:16 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:16 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:17 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:17 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:17 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:17 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:17 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:17 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:17 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:17 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:18 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:18 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:18 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:18 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:18 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:18 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:19 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:19 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:19 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:06:19 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:06:19 AM Debug:   AqualinkD: RS received packet of type Status length 12

**Reply by ballle98** [reply]
What I need to see in the logs is the equipment menu the is where the status is updated.  If it is not reaching this menu that is the problem.  If it is not reaching the equipment menu there should be some logs that are "Error:" and not "Debug:"

```
16:28:20.567 Debug:   PDA:       PDA Menu Line 0 =    EQUIPMENT    
16:28:20.569 Debug:   PDA:       PDA Menu Line 1 = FILTER PUMP  OFF
16:28:20.569 Debug:   PDA:       PDA Menu Line 2 = SPA          OFF
16:28:20.571 Debug:   PDA:       PDA Menu Line 3 = POOL HEAT    OFF
16:28:20.571 Debug:   PDA:       PDA Menu Line 4 = SPA HEAT     OFF
16:28:20.571 Debug:   PDA:       PDA Menu Line 5 = CLEANER      OFF
16:28:20.571 Debug:   PDA:       PDA Menu Line 6 = POOL LIGHT   OFF
16:28:20.573 Debug:   PDA:       PDA Menu Line 7 = SPA LIGHT    OFF
```

**Comment by heffneil** [comment]
I turned on my PDA to update my programing and the system doesn't show anything in the logs.  For better or worse I don't think I am saying connected between the pi and the panel?

**Reply by ballle98** [reply]
Might have to turn on the serial debug to see it

**Comment by heffneil** [comment]
May-04-22 08:07:42 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:07:42 AM Debug:   AqualinkD: RS received packet of type Lng Message length 24
May-04-22 08:07:42 AM Debug:   PDA:       To 0x61 of type      Lng Message | HEX: 0x10|0x02|0x61|0x04|0x40|0x20|0x20|0x20|0x20|0x20|0x57|0x45|0x44|0x20|0x20|0x38|0x3a|0x30|0x37|0x41|0x4d|0xde|0x10|0x03|
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 0 =      WED  8:07AM
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 1 = AIR             
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 2 =  68`            
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 3 = 
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 4 = POOL MODE    OFF
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 5 = POOL HEATER  OFF
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 6 = SPA MODE     OFF
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 7 = SPA HEATER   OFF
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 8 = MENU            
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu Line 9 = EQUIPMENT ON/OFF
May-04-22 08:07:42 AM Debug:   PDA:       PDA Menu highlighted line = 4 = POOL MODE    OFF
May-04-22 08:07:42 AM Debug:   PDA:       process_pda_packet_msg_long_time strptime Wed May  4 08:07:42 2022

May-04-22 08:07:42 AM Info:    PDA:       PDA Menu '2' Selectedline 'POOL MODE    OFF', Last line received '     WED  8:07AM'
May-04-22 08:07:42 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:07:42 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|
May-04-22 08:07:43 AM Debug:   AqualinkD: RS received packet of type Status length 12
May-04-22 08:07:43 AM Debug:   PDA:       To 0x61 of type           Status | HEX: 0x10|0x02|0x61|0x02|0x00|0x00|0x00|0x00|0x00|0x75|0x10|0x03|

**Comment by ballle98** [comment]
OK I posted some thoughts on #61

**Comment by heffneil** [comment]
So I was able to turn on my pump and spa- albeit very very slow and having to press the button multiple time to get it to respond but I was only able to turn the lights on but they will not turn off no matter what I do.  Not sure whats going on there and I have a fountain option I can't seem to operate within this program?

**Reply by ballle98** [reply]
It takes 5-10 seconds to complete the action.  There is a message that shows up in the banner that tells you it is processing the action.

![image](https://user-images.githubusercontent.com/2252524/179235801-871b4f41-fad8-4740-9d6f-194a03881356.png)

you can watch what is going on if you do logging in the foreground https://github.com/ballle98/AqualinkD/wiki#run-in-foreground-with-logging

Also note that actions are queued so you can turn multiple things on at once.  The first on/off will take 5-10 seconds but the subsequent ones will process much faster because the PDA is already on the equipment menu.

If you look through the logs you will see messages like this:

05:02:35.478 Debug:   Panl Prog: Thread 10,0xb5300f90 (Init PDA) finished in 17.473 sec
05:03:00.037 Debug:   Panl Prog: Thread 13,0xb5301cd0 (Switch PDA device on/off) finished in 6.012 sec
05:03:10.823 Debug:   Panl Prog: Thread 11,0xb5300f50 (Set SWG percent) finished in 7.909 sec
05:05:12.577 Debug:   Panl Prog: Thread 16,0xb5300e50 (SWG Boost) finished in 7.093 sec
